### ISSUE TITLE
timezones differ when reading and writing

### ISSUE THREAD

**Issue by D4edalus** [typeorm/typeorm#314]
mysql by default uses the systems timezone to store dates. when reading a sql date from the database it is correctly retrieved. the timezone of mysql and js Date match. however, when writing ( new Date() ) to the database i get a different result. the database corrects the datetime to a wrong value.

**Comment by pleerock** [comment]
can you please provide a real code example, values it writes and values it retrieves, and what is expected result

**Comment by D4edalus** [comment]
```
   let s : Screenshot = new Screenshot();
   s.id = 11111;
   s.time = new Date();
   console.log("Stored date local timezone " + s.time);
   console.log("Stored date utc+0 " + s.time.toISOString()); // Stored date local timezone Tue Feb 28 2017 21:52:47 GMT+0100 (Mitteleuropäische Zeit)
   await this.screenshotRepo.persist(s).catch(e => console.log(e)); // Stored date utc+0 2017-02-28T20:52:47.689Z

   let readss = await this.screenshotRepo.createQueryBuilder("ss").where("ss.id = 11111").getOne().catch(e => console.log(e));
   console.log("Read date local timezone " + readss.time); // Read date local timezone Tue Feb 28 2017 20:52:47 GMT+0100 (Mitteleuropäische Zeit)
   console.log("Read date utc+0 " + readss.time.toISOString()); // Read date utc+0 2017-02-28T19:52:47.000Z
```
As you can see one hour is wrongly subtracted from the time somewhere during conversion.

Definition of the time field looks like this:
@Column()
time: Date;

Phpmyadmin shows: 2017-02-28 20:51:28 ( afaik it is shown in local timezone (GMT+1) there )

**Comment by D4edalus** [comment]
so i dug a bit deeper ...
first i noticed typorm uses utc as a default
i should be able to set a column to use the local timezone with @Column({ localTimezone: true }) right? for me that does not change anything sadly

so next i figured that mixedDateToUtcDatetimeString and mixedDateToDatetimeString are used for datetime conversions. as utc seems to be default, i copied mixedDateToDatetimeString to mixedDateToUtcDatetimeString .. and the date conversion worked fine.

the time is stored correctly into the database with local and utc format. the problem seems to be reading from the database. the above conversion functions also dont seem to be called at all for reading a date from the database, they are only used for writing to it.

**Comment by pleerock** [comment]
Do you want to store your date in UTC?

**Comment by pleerock** [comment]
Take a look on [this test](https://github.com/typeorm/typeorm/blob/5fcc3aed2347b88163945214770ddddac28219a5/test/github-issues/176/issue-176.ts#L1) which shows how dates are working

**Comment by D4edalus** [comment]
i would prefer having (and it is i believe) UTC in the database.

https://github.com/typeorm/typeorm/blob/5fcc3aed2347b88163945214770ddddac28219a5/test/github-issues/176/issue-176.ts#L31
does not hold true for me, 2017-01-10 17:38:06 is visible in phpmyadmin ( = utc, good )

output after reading it:
Read date local timezone Tue Jan 10 2017 17:38:06 GMT+0100 (Mitteleuropäische Zeit)
Read date utc+0 2017-01-10T16:38:06.000Z
Read date utc+0 1484066286000

but i can understand the confusion now if this test is really working for you

**Comment by pleerock** [comment]
yes tests are passing on both CI and my local machine

**Comment by SergiSalvadorLozano** [comment]
Hi!

I seem to be encountering the same or a very similar issue. It looks like dates are being correctly stored in the database, but when they are retrieved the attributes are assigned an incorrect value (in my case, 1 hour is added. My local timezone is UTC+1).

For instance:

1. Create row with value: `{date_field: new Date("2017-03-10T15:45:42.000Z")}`
2. Check the database manually. The value has been correctly stored.
3. Retrieve the row that has just been created. Check the value: `{date_field: 2017-03-10T16:45:42.000Z}`

**Comment by D4edalus** [comment]
i didn't investigate further yet, but i'll pick up the project again soon. i suspect the testcases work when client and server are in the same timezone, but i havent verified that yet.

**Comment by SergiSalvadorLozano** [comment]
I had a temporary fix for my problem (typeorm was adding 1 hour to date fields on retrieval, so I corrected it manually subtracting it) which was working until a couple of weeks ago. However, with the change to Daylight Saving Time, it stopped working because 2 hours were being added instead of 1 (I went from GMT+1 to GMT+2).

The fix I have in place right now is:

`
let tzMinuteOffset = new Date().getTimezoneOffset();
myRetrievedDate.setTime(myRetrievedDate.getTime() + (1000 * 60 * tzMinuteOffset));
`

**Comment by D4edalus** [comment]
i'm unsure at what point this was fixed, but after upgrading to the latest 0.1.0-alpha.3 typeorm version it is working fine for me. the example ive posted above gives correct output now.

for future readers i also want to mention that using plain sql queries requires some manual time conversions. use SELECT `CONVERT_TZ(time,'localtime','UTC') FROM ...` to retrieve a date from the database. for comparisons to a stored date do not use `NOW()` but use its utc version `UTC_TIMESTAMP()`

**Comment by matkoklaic** [comment]
> Hi!
> 
> I seem to be encountering the same or a very similar issue. It looks like dates are being correctly stored in the database, but when they are retrieved the attributes are assigned an incorrect value (in my case, 1 hour is added. My local timezone is UTC+1).
> 
> For instance:
> 
> 1. Create row with value: `{date_field: new Date("2017-03-10T15:45:42.000Z")}`
> 2. Check the database manually. The value has been correctly stored.
> 3. Retrieve the row that has just been created. Check the value: `{date_field: 2017-03-10T16:45:42.000Z}`

I see the same problem. Don't know how to solve it.
